funnel/views/project.py

```diff
@@ -129,23 +132,81 @@ def csv(self):
+            initial_formdata = MultiDict(draft.body['form']) if draft is not None else None
+
+            # initialize forms with draft initial formdata.
+            # if no draft exists, initial_formdata is None. wtforms ignore formdata if it's None.
+            if self.obj.parent_project:
+                form = SubprojectForm(obj=self.obj, model=Project, formdata=initial_formdata)
+            else:
+                form = ProjectForm(obj=self.obj, parent=self.obj.profile, model=Project, formdata=initial_formdata)
+
+            if not self.obj.timezone:
+                form.timezone.data = current_app.config.get('TIMEZONE')
+
+            # if draft exists, add latest revision ID to the form
+            last_revision_id = draft.revision if draft is not None else None
+
+            return render_form(form=form, title=_("Edit project"), submit=_("Save changes"), autosave=True, last_revision_id=last_revision_id)
+        elif request.method == 'POST':
+            if getbool(request.args.get('form.autosave')):
+                if 'form.revision' not in request.form:
+                    # as form.autosave is true, the form should have `form.revision` field even if it's empty
+                    return {'error': _("Form must contain a valid revision ID.")}, 400
+
+                client_revision = request.form['form.revision']
+
+                # find the last draft
+                draft = Draft.query.get((Project.__tablename__, self.obj.uuid))
+
+                if draft is not None:
+                    if client_revision is None or (client_revision is not None and str(draft.revision) != client_revision):
+                        # draft exists, but the form did not send a revision ID,
+                        # OR revision ID sent by client does not match the last revision ID
+                        return {'error': _("There has been changes to this draft since you last edited it. Please reload.")}, 400
+                    elif client_revision is not None and str(draft.revision) == client_revision:
+                        # revision ID sent my client matches, save updated draft data and update revision ID
+                        existing = MultiDict(draft.body['form'])
+                        incoming = MultiDict(request.form.items(multi=True))
+                        for key in incoming.keys():
+                            if existing[key] != incoming[key]:
+                                existing[key] = incoming[key]
+                        draft.body = {'form': existing}
+                        draft.revision = uuid4()
+                else:
+                    # no draft exists, create one
+                    draft = Draft(
+                        table=Project.__tablename__, table_row_id=self.obj.uuid,
+                        body={'form': request.form.items(multi=True)}, revision=uuid4()
+                        )
+                db.session.add(draft)
```

Comment:
This requires a CSRF check. Without the check, a malicious third party website can dump junk content into the Draft model by performing a cross-site POST (which is what CSRF protects from).

1. Do a CSRF check using `Form().validate_on_submit()` (the base Form object has nothing but a CSRF field).
2. Remove the CSRF field when saving to the Draft model. It will cause problems when restored later.